### DISCUSSION TITLE
Make flux install additive when there's an existing configuration

### DISCUSSION THREAD

**Discussion by squaremo** [fluxcd/flux2#852] [Proposals] [answered]
`flux install` is idempotent, meaning that running exactly the same command several times consecutively will give the same state each time.

But idempotence is not always the most useful property; for example, if I wish to _add_ controllers to a Flux v2 configuration, I have to know the original command that was run, otherwise (if I just mention the new things) it will remove components, or (if I put the output in a new file) it will duplicate some common parts, leading to an invalid configuration.

Proposal: let people provide an existing configuration to `flux install`, and produce a new configuration including the existing pieces and the new pieces as specified. This may only be feasible when dealing with files, i.e., `--export` is used. As a sketch:

```bash
$ cat gotk-components.yaml | flux install --add --components=image-reflector-controller --export > gotk-component.yaml
# ...
$ diff gotk-components.yaml
# diff output with lines for image-reflector-controller deployment and CRD added
```

kyaml functions work like this, so perhaps a kyaml pipeline will help.

**Comment by stefanprodan** [comment, marked as answer]
> I have to know the original command that was run

A different solution was proposed here https://github.com/fluxcd/flux2/discussions/304

@squaremo I don't think your solution is feasible due to:
* controllers at different versions are not guarded to work together unless that exact combination has passed e2e (e.g. source-controller/api 0.2 is not compatible with helm-controller 0.7)
* RBAC changes require all controller's SA to be present, adding a controller with SA to a setup created with flux <0.7 will break 
* how are we going to deal with conflicts regarding log level, private container registry, image pull secret and all the others customisations that right now apply to all components

**Reply by squaremo** [reply]
Merging the existing and the new configuration would work well enough for e..g, versioning. You are right about things like log level and private container registries; not recording that somewhere makes this approach pretty tricky. (Of course, recording it somewhere brings its own problems! Computer.)
